src/p384/address.rs

```diff
@@ -0,0 +1,215 @@
+        self.as_bytes()[0] == Self::REQUIRED_PREFIX
+    }
+
+    pub fn write_to_string(&self, s: &mut String, prefix: bool) {
+        if prefix {
+            s.push_str(PREFIX_ADDRESS);
+        }
+        first_128_to_string(&self.as_bytes()[..16], s);
+        s.push('.');
+        base62::encode_8to11(u64::from_be(self.0[2]), s);
+        base62::encode_8to11(u64::from_be(self.0[3]), s);
+        base62::encode_8to11(u64::from_be(self.0[4]), s);
+        base62::encode_8to11(u64::from_be(self.0[5]), s);
+    }
+}
+
+impl TryFrom<[u8; 48]> for Address {
+    type Error = InvalidParameterError;
+
+    #[inline]
+    fn try_from(value: [u8; 48]) -> Result<Self, Self::Error> {
+        let a = Self(unsafe { transmute(value) });
+        if a.is_valid() {
+            Ok(a)
+        } else {
+            Err(ADDRESS_ERR)
+        }
+    }
+}
+
+impl From<Address> for [u8; 48] {
+    #[inline(always)]
+    fn from(value: Address) -> Self {
+        unsafe { transmute(value) }
+    }
+}
+
+impl ToFromBytes for Address {
+    #[inline]
+    fn read_bytes<R: std::io::Read>(r: &mut R) -> std::io::Result<Self> {
+        let mut tmp = Self([0; 6]);
+        r.read_exact(tmp.as_mut_bytes())?;
+        if tmp.is_valid() {
+            Ok(tmp)
+        } else {
+            Err(std::io::Error::new(std::io::ErrorKind::Other, ADDRESS_ERR.0))
+        }
+    }
+
+    #[inline(always)]
+    fn write_bytes<W: std::io::Write>(&self, w: &mut W) -> std::io::Result<()> {
+        w.write_all(self.as_bytes())
+    }
+}
+
+impl ToString for Address {
+    fn to_string(&self) -> String {
+        let mut s = String::with_capacity(Self::STRING_SIZE);
+        self.write_to_string(&mut s, true);
+        s
+    }
+}
+
+impl Debug for Address {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.to_string().as_str())
+    }
+}
+
+impl FromStr for Address {
+    type Err = InvalidParameterError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s = s.trim();
+        let s = s.strip_prefix(PREFIX_ADDRESS).unwrap_or(s);
+        let sb = s.as_bytes();
+        if sb.len() == Self::STRING_SIZE_NO_PREFIX && sb[31] == b'.' {
+            let prefix = ShortAddress::from_str(&s[..31])?;
+            Ok(Address([
+                prefix.0[0],
+                prefix.0[1],
+                base62::decode_11to8(&sb[32..43])?.to_be(),
+                base62::decode_11to8(&sb[43..54])?.to_be(),
+                base62::decode_11to8(&sb[54..65])?.to_be(),
+                base62::decode_11to8(&sb[65..76])?.to_be(),
+            ]))
+        } else {
+            Err(ADDRESS_ERR)
+        }
+    }
+}
+
+impl AsRef<[u8]> for Address {
+    #[inline(always)]
+    fn as_ref(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+impl Hash for Address {
+    #[inline(always)]
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write_usize(self.0[0] as usize)
```

Comment:
Documenting the discussion here for posterity:
So that is code I didn't actually change in this PR, but just moved into a file.
Reading it it is using the first 64 bits of the Address as the hashed value, while the required prefix of 0xfc is only 8 bits long.
So the hash function is getting 56 bits of entropy from this implementation.
I believe Adam wanted addresses to be fast to hash for hashtables.